dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/ReferenceBeanManager.java

```diff
@@ -0,0 +1,253 @@
+     *
+     * @param referenceBean
+     * @throws Exception
+     */
+    private void initReferenceBean(ReferenceBean referenceBean, boolean canCreateProxy) throws Exception {
+
+        if (referenceBean.getReferenceConfig() != null) {
+            return;
+        }
+
+        Environment environment = applicationContext.getEnvironment();
+        Map<String, Object> referenceProps = referenceBean.getReferenceProps();
+        if (referenceProps == null) {
+            MutablePropertyValues propertyValues = referenceBean.getPropertyValues();
+            if (propertyValues == null) {
+                throw new RuntimeException("ReferenceBean is invalid, missing 'propertyValues'");
+            }
+            referenceProps = toReferenceProps(propertyValues, environment);
+        }
+
+        //resolve placeholders
+        resolvePlaceholders(referenceProps, environment);
+
+        //create real ReferenceConfig
+        ReferenceConfig referenceConfig = ReferenceBeanBuilder.create(new AnnotationAttributes(new LinkedHashMap<>(referenceProps)), applicationContext)
+                .defaultInterfaceClass(referenceBean.getObjectType())
+                .build();
+
+        referenceBean.setReferenceConfig(referenceConfig);
+
+        // register ReferenceConfig
+        DubboBootstrap.getInstance().reference(referenceConfig);
+
+        //TODO add after DubboBootstrap is started
+        if (canCreateProxy && referenceConfig.shouldInit()) {
+            ReferenceConfigCache.getCache().get(referenceConfig);
+        }
+    }
+
+    private void resolvePlaceholders(Map<String, Object> referenceProps, PropertyResolver propertyResolver) {
+        for (Map.Entry<String, Object> entry : referenceProps.entrySet()) {
+            Object value = entry.getValue();
+            if (value instanceof String) {
+                String valueToResovle = (String) value;
+                entry.setValue(propertyResolver.resolveRequiredPlaceholders(valueToResovle));
+            } else if (value instanceof String[]) {
+                String[] strings = (String[]) value;
+                for (int i = 0; i < strings.length; i++) {
+                    strings[i] = propertyResolver.resolveRequiredPlaceholders(strings[i]);
+                }
+                entry.setValue(strings);
+            }
+        }
+    }
+
+    private Map<String, Object> toReferenceProps(MutablePropertyValues propertyValues, PropertyResolver propertyResolver) {
+        Map<String, Object> referenceProps;
+        referenceProps = new LinkedHashMap<>();
+        for (PropertyValue propertyValue : propertyValues.getPropertyValueList()) {
+            String propertyName = propertyValue.getName();
+            Object value = propertyValue.getValue();
+            if ("methods".equals(propertyName)) {
+                ManagedList managedList = (ManagedList) value;
+                List<MethodConfig> methodConfigs = new ArrayList<>();
```

Comment:
`LinkedList` would be more efficient in this sutiation

Reply:
ArrayList is faster when adding fewer elements, generally there is no need to expand the capacity, and toArray() is also more efficient.